firebase-functions/functions/src/operations/playerOperations.ts

```diff
@@ -1,32 +1,57 @@
-﻿import { getPlayersToDeactivate } from './choiceOperations';
-import * as admin from 'firebase-admin';
+﻿import * as admin from 'firebase-admin';
 import { firestore } from 'firebase-admin';
 import { collections } from '../constants/collections';
 import { Player } from '../models/player';
 import DocumentData = firestore.DocumentData;
 import QueryDocumentSnapshot = firestore.QueryDocumentSnapshot;
+import { Game } from '../models/game';
+import { GameType } from '../models/game-type';
+import { Result } from '../models/result';
+import { getCurrentGameType } from './roomOperations';
+import { addLastOneActiveToGame } from './GameOperations';
 
-export async function deactivatePlayers(roomId: string, initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[]): Promise<void> {
-  const playersToDeactivate = await getPlayersToDeactivate(initiallyActivePlayers, roomId);
+export async function deactivatePlayers(roomId: string, game: Game, initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[]): Promise<void> {
+  const playersToDeactivate = await getPlayersToDeactivate(game, roomId);
   if (!playersToDeactivate.length) {
     return;
   }
   await updatePlayers(playersToDeactivate, { isObserver: true });
   await determineLastOneActive(initiallyActivePlayers, playersToDeactivate, roomId);
 }
 
+export async function getActivePlayersInRoom(roomId: string): Promise<QueryDocumentSnapshot<DocumentData>[]> {
+  return (await admin.firestore().collection(collections.players).where('room', '==', roomId).get())
+    .docs.filter(playerDoc => !(playerDoc.data() as Player).isObserver);
+}
+
 export async function resetPlayerChoices(players: QueryDocumentSnapshot<DocumentData>[]): Promise<void> {
-  await updatePlayers(players, { choice: null });
+  await updatePlayersDoc(players, { choice: null });
 }
 
-async function determineLastOneActive(initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[], playersToDeactivate: QueryDocumentSnapshot<DocumentData>[], roomId: string): Promise<void> {
-  const stillActivePlayers = initiallyActivePlayers.filter(player => !playersToDeactivate.includes(player));
-  if (stillActivePlayers.length === 1) {
-    await admin.firestore().collection(collections.rooms).doc(roomId).update({ lastOneActive: stillActivePlayers[0].id });
-    await updatePlayers(stillActivePlayers, { isObserver: true });
+async function determineLastOneActive(initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[], playersToDeactivate: string[], roomId: string): Promise<void> {
+  if (initiallyActivePlayers.length === playersToDeactivate.length + 1) {
+    const lastOneActive = initiallyActivePlayers.find(player => !playersToDeactivate.includes(player.id));
+    if (lastOneActive) {
+      await addLastOneActiveToGame(lastOneActive.id, roomId);
```

Comment:
Is this check necessary, or does it maybe even hide a bug? 
If we get to that line and we didn't find a lastOneActive, something is wrong I think.

Reply:
Logically, it won't happen but the method `find `by definition return undefined that should be guarded. Anyway, I removed the guard to another level.